cmd/snap-bootstrap/cmd_scan_disk.go

```diff
@@ -0,0 +1,259 @@
+	return scanDisk()
+}
+
+type Partition struct {
+	Name string
+	UUID string
+}
+
+func getProbeValue(probe C.blkid_probe, name string) (string, err) {
+	var value *C.char
+	var value_len C.size_t
+	entryname := C.CString(name)
+	defer C.free(unsafe.Pointer(entryname))
+	res := C.blkid_probe_lookup_value(probe, entryname, &value, &value_len)
+	if res < 0 {
+		return "", fmt.Errorf("Probe value was not found: %s", name)
+	}
+	if value_len > 0 {
+		return C.GoStringN(value, C.int(value_len-1))
+	} else {
+		return "", fmt.Errorf("Probe value has unexpected size")
+	}
+}
+
+func isGpt(probe C.blkid_probe) bool {
+	pttype, err := getProbeValue(probe, "PTTYPE")
+	if err != nil {
+		return false
+	}
+	return pttype == "gpt"
+}
+
+func probePartitions(node string) ([]Partition, error) {
+	cnode := C.CString(node)
+	defer C.free(unsafe.Pointer(cnode))
+	probe, err := C.blkid_new_probe_from_filename(cnode)
+	if probe == nil {
+		return nil, err
+	}
+	defer C.blkid_free_probe(probe)
+
+	C.blkid_probe_enable_partitions(probe, 1)
+	C.blkid_probe_set_partitions_flags(probe, C.BLKID_PARTS_ENTRY_DETAILS)
+	C.blkid_probe_enable_superblocks(probe, 1)
+
+	res, err := C.blkid_do_safeprobe(probe)
+	if res < 0 {
+		return nil, err
+	}
+
+	if !isGpt(probe) {
+		return nil, nil
+	}
+
+	partitions, err := C.blkid_probe_get_partitions(probe)
+	if partitions == nil {
+		return nil, err
+	}
+
+	npartitions := C.blkid_partlist_numof_partitions(partitions)
+	ret := make([]Partition, 0)
+	for i := 0; i < int(npartitions); i++ {
+		partition := C.blkid_partlist_get_partition(partitions, C.int(i))
+		label := C.GoString(C.blkid_partition_get_name(partition))
+		uuid := C.GoString(C.blkid_partition_get_uuid(partition))
+		fmt.Fprintf(os.Stderr, "Found partition %s %s\n", label, uuid)
+		ret = append(ret, Partition{label, uuid})
+	}
+
+	return ret, nil
+}
+
+func scanDiskNode(node string) {
+	fmt.Fprintf(os.Stderr, "Scanning disk %s\n", node)
+	fallback := false
+	bootUUID, _, err := efi.ReadVarString("LoaderDevicePartUUID-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "No efi var, falling back: %s\n", err)
+		fallback = true
+	} else {
+		bootUUID = strings.ToLower(bootUUID)
+	}
+
+	partitions, err := probePartitions(node)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Cannot get partitions: %s\n", err)
+		return
+	}
+
+	found := false
+	has_seed := false
+	var seed_uuid string
+	has_boot := false
+	var boot_uuid string
+	has_data := false
+	var data_uuid string
+	has_save := false
+	var save_uuid string
+	for _, part := range partitions {
+		if !fallback {
+			if part.UUID == bootUUID {
+				found = true
+			}
+		}
+		if part.Name == "ubuntu-seed" {
+			has_seed = true
+			seed_uuid = part.UUID
+		} else if part.Name == "ubuntu-boot" {
+			has_boot = true
+			boot_uuid = part.UUID
+		} else if part.Name == "ubuntu-data-enc" {
+			has_data = true
+			data_uuid = part.UUID
+		} else if part.Name == "ubuntu-data" {
```

Comment:
if someone appends unencrypted ubuntu-data to a hard-drive, then it will be mounted instead of ubuntu-data-enc? surely we should try to find -enc paritions, and not search for non-encrypted ones afterwords?

Also I thought udev database already has all of this information inside it's database, which maybe can be queried directly in the udev rule alone without this helper binary? Apart from the LoaderDevicePartUUID bit, which i thought there should be a udev symlink for.

Reply:
> if someone appends unencrypted ubuntu-data to a hard-drive, then it will be mounted instead of ubuntu-data-enc? surely we should try to find -enc paritions, and not search for non-encrypted ones afterwords?

Maybe I mixed up, but I think the filesystem labels have "-enc" suffix. But the partition name does not. See for example:

```
/dev/vda4: UUID="2eeba654-032f-4384-8a46-075be2b415b5" LABEL="ubuntu-save-enc" TYPE="crypto_LUKS" PARTLABEL="ubuntu-save" PARTUUID="a154643d-4398-714a-9704-71335e5ea856"
```

`PARTLABEL` is not the same as `LABEL`

Reply:
Inside ubuntu-save-enc there is ubuntu-save. Outside of ubuntu-save-enc there can be a malicious second partition with an FS ubuntu-save. Does above code handle this fine? As far as I can tell, after scanning and finding ubuntu-data-enc and using it; it will then find unencrypted ubuntu-data and override it.

Somehow this feels odd that after the first one is detected, it can be detected again, and override previously found result.

Reply:
> Also I thought udev database already has all of this information inside it's database, which maybe can be queried directly in the udev rule alone without this helper binary? Apart from the LoaderDevicePartUUID bit, which i thought there should be a udev symlink for.

Good question. I have tried before to use more rules and less code. But there were issues. And I do not remember. I will try again to document why this is not possible.

I am guessing it was the fallback for non-uefi boot that was a bit tricky.

`gpt-auto-generator` can create a symlink. But it has to be a partition with a specific uuid on the same disk as the booting partition. We do not use that uuid. And we should not use it. It is not made for that.

Reply:
> Inside ubuntu-save-enc there is ubuntu-save. Outside of ubuntu-save-enc there can be a malicious second partition with an FS ubuntu-save.

I am not sure I understood this.

This binary only looks at partition names. Not filesystem labels. If there are other partitions with conflicting filesystem labels, we will not look at them, because we only look at the partition names.

Reply:
Today, someone can remove an encrypted ubuntu-save and replace it by a non encrypted one. `snap-bootstrap` will then fallback to it. What would be the difference?

Reply:
> > Also I thought udev database already has all of this information inside it's database, which maybe can be queried directly in the udev rule alone without this helper binary? Apart from the LoaderDevicePartUUID bit, which i thought there should be a udev symlink for.
> 
> Good question. I have tried before to use more rules and less code. But there were issues. And I do not remember. I will try again to document why this is not possible.

Now I remember. It is possible to not call it on `ENV{DEVTYPE}=="partition"`. Though we need to make sure that `IMPORT{builtin}="blkid [...]"` has been called yet, so post `60-persistent-storage.rules`. Then what we have to do is compare `ID_PART_ENTRY_UUID` with `UBUNTU_SEED` (or the 3 others).

However, for `ENV{DEVTYPE}=="disk"` we do need to scan the partitions and find the booting one. And this we cannot do with just udev rules. We have `LoaderDevicePartUUID` which points to the boot partition. But we are processing the disk at this point.

Reply:
> `gpt-auto-generator` can create a symlink.

By the way, I was wrong about it. It is in an udev rule that the symlink is created. `gpt-auto-generator` only makes `sysroot.mount` and `-.mount` to use that symlink.